### ISSUE TITLE
FileSelector dropdown menu too narrow

### ISSUE THREAD

**Issue by philippjfr** [ioam/paramnb#4]
When using the FileSelector the widget itself is wide enough however when clicking on the dropdown you basically can't read filenames over a certain length. This is using the ipywidgets 5.1.5.

![image](https://cloud.githubusercontent.com/assets/1550771/15855119/a79c0b4c-2ca5-11e6-989f-32484fbec05a.png)

**Comment by jbednar** [comment]
D'oh!  That's annoying.  I don't know if there is a parameter to control that somewhere...

**Comment by jbednar** [comment]
Alternatively, there is a mapping between the textual description and the item being selected, so maybe we should remove any common prefix from the dropdown?

**Comment by philippjfr** [comment]
> Alternatively, there is a mapping between the textual description and the item being selected, so maybe we should remove any common prefix from the dropdown?

That sounds like a sensible thing to do independent of this issue and will at least ameliorate the need for  a proper fix for this issue.

**Comment by jbednar** [comment]
Hmm; I'm now not so sure it's sensible.  I implemented it in a few lines:

``` python
from os.path import commonprefix, dirname, sep

class FileSelector(param.ObjectSelector):
    ...

    def get_range(self):
        named_objs = super(FileSelector, self).get_range()
        prefix = commonprefix([dirname(name)+sep for name in named_objs.keys()])
        return {name[len(prefix):]:obj for name,obj in named_objs.iteritems()}
```

and it works:

![image](https://cloud.githubusercontent.com/assets/1695496/15911648/57e2f5ce-2d96-11e6-92dc-86ee871f4d21.png)

but it's not showing that the `\_\_init\_\_.py` file it found is actually in `paramnb/`, not any of the other directories it could have been in.  It's probably ok not to show that in the dropdown, but then in the actual widget that's left, presumably it's important, and there's no way for the user to see that information.  It's not _that_ important, since by definition it's the only one that matched the glob pattern (or there would have been a shorter common prefix), but still this seems confusing.  I can't see any obvious way around it...

**Comment by philippjfr** [comment]
Yes, that's not ideal, I could imagine updating the original query with the common prefix combined with the part of the glob that was not unique or is concerned with the filename and extension. The logic could become involved though.

**Comment by jbednar** [comment]
I.e., change the string `../*/*.py*` to `../paramnb/*.py*`, because paramnb is the only directory containing *.py*?  That would solve this problem, but seems very confusing for the users.

Oddly, the basic Dropdown doesn't seem to have this truncation problem:

![image](https://cloud.githubusercontent.com/assets/1695496/15912071/fd3eeb08-2d97-11e6-8b9a-1e68410ed383.png)

So it must be due to how Dropdown is being used for FileSelector?

**Comment by philippjfr** [comment]
> That would solve this problem, but seems very confusing for the users.

True, especially if it's happening as you're typing, globs would magically autocomplete.

> So it must be due to how Dropdown is being used for FileSelector?

Sounds like it, hopefully we can avoid dropping the prefix then.

**Comment by jbednar** [comment]
I guess there is a hybrid approach: strip the common prefix if it's in the path spec as text, but don't do it if it's a glob.  I think that's what will make the most sense to users, because repeating that common string looks ridiculous, but in the glob case it's not repeated, it's novel information compared to the glob.  I'll have to see if that can be expressed clearly and reliably.

Meanwhile, yes, hopefully there's a way to find out where the dropdown cropping is happening, but it's a fairly mysterious problem.

**Comment by jbednar** [comment]
It looks like it's easy to achieve this modified prefix suppression, by simply adding the path spec to the list from which the common prefix is found.  That successfully removes the common prefix if it's spelled out:

![image](https://cloud.githubusercontent.com/assets/1695496/15912530/985e4e74-2d9a-11e6-8ac8-1b5813e75586.png)

but not if it isn't:

![image](https://cloud.githubusercontent.com/assets/1695496/15912554/b53167fc-2d9a-11e6-8607-302430172271.png)

Now the common prefix suppression seems like sensible behavior, to me.

**Comment by philippjfr** [comment]
Yes, that seems like the most sensible behavior, no redundant or missing information.

**Comment by jbednar** [comment]
Prefix suppression now committed and pushed, 8d098a4.  I'll leave this open, as I still don't want the width of the field to be limited in this way, but in practice this problem should now be rare.

**Comment by philippjfr** [comment]
Great, looks good. I'll try to have a look at this issue at some point, but as you say, it's not a major priority now.

**Comment by philippjfr** [comment]
This seems to at least make the width of the dropdown menu consistent with the menu itself. Not sure how best to include it though:

``` css
.widget-dropdown .dropdown-menu {
   width: 100%
}
```

**Comment by jbednar** [comment]
Ok, I'll add that.  Thanks!

**Comment by jbednar** [comment]
Fixed in 87d2c913.

**Comment by jbednar** [comment]
One remaining issue: if a short filename is selected, but the dropdown has much longer filenames, the dropdown will only be as wide as the widget, not as the longest filename.  A workaround is to select one of the longer filenames first, which is awkward but at least does reveal the information.  It would be good to make the dropdown expand in this case.

**Comment by jbednar** [comment]
Another remaining issue: The selected item, if very long, is truncated:

![image](https://cloud.githubusercontent.com/assets/1695496/23386432/07504b9c-fd1c-11e6-9f33-2c88f6f57812.png)

Here the longnames directory has three files, each with the same first 30 chars but differing after that, yet the selected one is truncated shorter than the others.
